### PR TITLE
Gitlab protocol report: gitaly node failure info

### PR DESCRIPTION
**Pull request by ivanbukhtiyarov**
For https://github.com/cqfn/degitx/issues/179 - added an info about sqlElector working and recovery in case of node failure

### CODE REVIEW

File: reports/gitlab-commit-protocol/gitlab-commit-protocol.tex [hunk 1]
```diff
@@ -98,7 +98,34 @@ \section{3PC}
...
+shard). It enables multiple, redundant Praefect processes to run,
+which is needed to eliminate a single point of failure in Gitaly High
+Avaiability.
+
+The sqlElector is responsible for:
+\begin{itemize}
+\item Monitoring and updating the status of all nodes within the shard.
+\item Electing a new primary of the shard based on the health.
+\end{itemize}
+Also very Praefect node periodically performs a health check RPC with a Gitaly node.  Once the health checks are complete, Praefect node does a \textit{SELECT} from \textit{node\_status} table to determine healthy nodes.
```
Comment (IlyaMoskva) [inline reports/gitlab-commit-protocol/gitlab-commit-protocol.tex:112]: very - every?

File: reports/gitlab-commit-protocol/gitlab-commit-protocol.tex [hunk 2]
```diff
@@ -98,7 +98,34 @@ \section{3PC}
...
+The primary of each shard is listed in the
+\textit{shard\_primaries}. If the current primary is in the healthy
+node list, then \textit{sqlElector} updates its internal state to match.
+
+Otherwise, if there is no primary or it is unhealthy, any Praefect node
+can elect a new primary by choosing candidate from the healthy node
+list. If there are no candidate nodes, the primary is demoted by setting the \textit{demoted} flag in \textit{shard\_primaries}.
+
+In case of a failover, the virtual storage is marked as read-only until writes are manually enabled
+again. This status is stored in the \textit{shard\_primaries} table's \textit{read\_only }column. If it is set, mutator RPCs against the storage shard should be blocked in order to prevent new primary from
```
Comment (IlyaMoskva) [inline reports/gitlab-commit-protocol/gitlab-commit-protocol.tex:123]: Please, rewrite this sentence to be unambiguous. Too long and complex to understand right.

File: reports/gitlab-commit-protocol/gitlab-commit-protocol.tex [hunk 3]
```diff
@@ -98,7 +98,34 @@ \section{3PC}
...
+diverging from the previous primary before data recovery attempts have been made.
+
+The different scenarios of node failures described \emph{\href{https://gitlab.com/gitlab-org/gitaly/-/blob/master/internal/praefect/nodes/sql_elector.go\#L401}{in the code comments}}.
+
+\todo{figure out the process of prefect (TM) backup}
```
Comment (IlyaMoskva) [inline reports/gitlab-commit-protocol/gitlab-commit-protocol.tex:128]: pr**a**efect

File: reports/gitlab-commit-protocol/gitlab-commit-protocol.tex [hunk 4]
```diff
@@ -98,7 +98,34 @@ \section{3PC}
...
+The sqlElector is responsible for:
+\begin{itemize}
+\item Monitoring and updating the status of all nodes within the shard.
+\item Electing a new primary of the shard based on the health.
+\end{itemize}
+Also very Praefect node periodically performs a health check RPC with a Gitaly node.  Once the health checks are complete, Praefect node does a \textit{SELECT} from \textit{node\_status} table to determine healthy nodes.
+
+The primary of each shard is listed in the
+\textit{shard\_primaries}. If the current primary is in the healthy
+node list, then \textit{sqlElector} updates its internal state to match.
```
Comment (IlyaMoskva) [inline reports/gitlab-commit-protocol/gitlab-commit-protocol.tex:116]: What kind of list - ordered by some sign or not? How system choose which node should be taken from it?

**Review by IlyaMoskva — CHANGES_REQUESTED** [review]
@ivanbukhtiyarov Thanks for the clarification. Please, look at my comments.

**Comment by ivanbukhtiyarov** [comment]
Thanks for review @IlyaMoskva, fixed mistakes and tried to clarify primary electing strategy.

**Review by IlyaMoskva — APPROVED** [review]
Great job, @ivanbukhtiyarov

**Review by g4s8 — APPROVED** [review]
@ivanbukhtiyarov thanks

**Pull request merged**